src/java/org/apache/cassandra/utils/FastByteOperations.java

```diff
@@ -331,11 +348,17 @@ public static int compareTo(ByteBuffer buffer1, ByteBuffer buffer2)
                 obj1 = buffer1.array();
                 offset1 = BYTE_ARRAY_BASE_OFFSET + buffer1.arrayOffset();
             }
+            else if (buffer1.isReadOnly() && !buffer1.isDirect())
+            {
+                obj1 = theUnsafe.getObject(buffer1, HEAP_HB_FIELD_OFFSET);
+                offset1 = BYTE_ARRAY_BASE_OFFSET;
```

Comment:
is this actually true?  what about the following

```
ByteBuffer.wrap(new byte[42]).position(10).asReadOnly()
```

In this case won't the offset be `BYTE_ARRAY_BASE_OFFSET + buffer1.arrayOffset()`?